pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -77,16 +81,88 @@ func NewClusterMemberRemovalController(
 			masterNodeInformer,
 			masterMachineInformer,
 			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
 		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
 }
 
 func (c *clusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
-	// stop if the machine API is not functional
+	var errs []error
+
+	if err := c.removeMemberWithoutMachine(ctx); err != nil {
+		errs = append(errs, err)
+	}
+
+	// only attempt to scale down if the machine API is functional
 	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
-		return err
+		errs = append(errs, err)
+		return kerrors.NewAggregate(errs)
 	} else if !isFunctional {
 		return nil
 	}
+
+	if err := c.attemptToScaleDown(ctx); err != nil {
+		errs = append(errs, err)
+	}
+
+	return kerrors.NewAggregate(errs)
+}
+
+// attemptToScaleDown attempts to remove the member only once we have identified that
+// a Machine resource is being deleted and a replacement member has been created
+func (c *clusterMemberRemovalController) attemptToScaleDown(ctx context.Context) error {
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	memberMachines := ceohelpers.FilterMachinesWithMachineDeletionHook(masterMachines)
```

Comment:
When deciding whether or not to scale-down a voting member this count should only be voting member machines.

Otherwise given `desiredControlPlaneReplicasCount = 3` and if we have `3 voting + 1 learner` members, then `len(memberMachines) = 4` since learner member machines also have a deletion hook.
Deleting the voting member would trigger a scale-down since  `len(memberMachines) > desiredControlPlaneReplicasCount` but now we've just replaced a voting member with a learner member (which may not get promoted).

So we also need to filter out the learner members here and change `memberMachines => votingMemberMachines` when considering scaling down a voting machine.

Reply:
With the above you would still need to handle the cleanup of learner members pending deletion. That does not depend on `len(memberMachines) <= desiredControlPlaneReplicasCount`, but rather we remove the learner member whenever its machine is pending deletion.

I would suggest moving the learner member removal logic to its own function to make it simpler, but your call on addressing that wherever makes sense.

Reply:
yes, you are right, I have updated the PR, please see if you like it.